### DISCUSSION TITLE
Shifting of painting after complete drawing

### DISCUSSION THREAD

**Discussion by Chris2137** [hm21/pro_image_editor#51] [Q&A] [answered]
Hi, it is a follow up on https://github.com/hm21/pro_image_editor/issues/50

I tried with the latest version 3.0.0-dev.9, the shifting improved a lot. But when the drawing is away from the center, the shifting is obvious, here is what I draw in the picture 
original : 
![image](https://github.com/hm21/pro_image_editor/assets/22078631/f72e07ea-dc0a-45fe-bf49-7816e471050c)

draw a red line around the lid :
![image](https://github.com/hm21/pro_image_editor/assets/22078631/a02554dd-fff8-478b-89a8-3233072f6c25)

completed editing : 
![image](https://github.com/hm21/pro_image_editor/assets/22078631/07faea20-46fb-4fb1-85a8-3aa102b69e19)

Open up painting mode again : 
![image](https://github.com/hm21/pro_image_editor/assets/22078631/c5c3943c-cce0-4355-aa83-613024d1e19c)

Is it going to be fixed? Thanks.

btw, the painting always start after some initial movement, making it hard to start drawing at precise location. Hopefully it can be fixed too. Many thanks.

**Comment by hm21** [comment]
Thank you for the detailed information. The problem by the image is just by the creation (after you close the painting editor the first time), but after you open/close a subeditor the position is correct, right? 
Btw, I add both issues to my to-do list [here](https://github.com/users/hm21/projects/5/views/1?filterQuery=), so I will give you feedback here after I had time to fix it.

**Reply by Chris2137** [reply]
Sadly no. To re-iterate the problem, 
when I use previous version, 2.7.9, 
the situation is : after I edit inside painting editor and close it, it shows correctly on screen; when I open the editor again, it shifts to the left (as discussed in https://github.com/hm21/pro_image_editor/issues/50). But even if I repeatedly edit the image, it always show correctly on the final image.

In the dev version, 3.0.0-dev
the situation becomes : after I edit inside painting editor and close it, it shows **incorrectly** on screen (drawing float above the lid); when I open the editor again, it shows same as the 'final image' that the painting is shifted upward. Means I lost my original drawing location. It is shifted upward no matter inside or outside of the editor.  Only if my drawing is in the middle of the picture, there will be no shifting happen.

Hope it is clear for you.

**Reply by Chris2137** [reply]
To supplement, in 2.7.9, I got the correct final image, the only problem is when I open editor again, my drawing is shifted.
In 3.0.0-dev, I can't get correct final image, the drawing is shifted both in final image and when re-open editor (to be fair, the shifting is reduced comparing with previous version)

**Reply by hm21** [reply, marked as answer]
Okay, I did some tests and we have two options now.

## Option 1:
To recalculate the layer position, I can add a function inside the ProImageEditorConfigs that allows you to recalculate the position depending on the size of your main editor. It will look like this:
```dart
 configs: ProImageEditorConfigs(
  test: (List<Offset?> offsetPositions, Size paintEditorSize, Size mainEditorSize) {
    List<Offset?> updatedOffsetPositions;

    /// recalculate offset position

    return updatedOffsetPositions;
  }
)
```
In the real function I will also add stroke width and all stuff because you need next to update all offset positions also to scale down the layer. In my opinion, this will be a bit hard because in “freestyle” mode the layer has a lot of offset positions.

## Option 2: 
We open the subeditor in the same size and position as the main editor, as you asked me before. The downside here is that any content outside the subeditor will not be interactive. Below I'll post an example where I add a barrierColor that the user can see it's disabled. When the user clicks on this area, the dialog will close and he can interact again. 

![image](https://github.com/hm21/pro_image_editor/assets/40503456/1fd294ff-34e2-406b-a0a5-50071125faed)

**Reply by Chris2137** [reply]
Option 2 would be perfect for my case. The content outside not interactive is perfectly fine in my opinion.
Look forward to the fix. Thanks a lot @hm21

**Reply by hm21** [reply]
Okay, I added option 2 in the dev branch `3.0.0-dev.12`. You can now set it up that way:

```dart
ProImageEditor.asset(
      'demo.jpg',
      configs: ProImageEditorConfigs(
            imageEditorTheme: ImageEditorTheme(
                subEditorPage: SubEditorPageTheme(
                    enforceSizeFromMainEditor: true,
                    positionTop: 0,
                    positionLeft: 0,
                    barrierDismissible: true,
                    barrierColor: const Color(0x90272727),
                    borderRadius: BorderRadius.circular(10),
                    transitionsBuilder: (context, animation, secondaryAnimation, child) {
                      return child;
                    }),
             ),
      ),
);
```

**Reply by Chris2137** [reply]
Before opening subeditor
![image](https://github.com/hm21/pro_image_editor/assets/22078631/9f563039-11db-468f-9472-687e1db23fe2)

After opening subeditor (the subeditor is a bit shifted upwards)
I believe it is because you're respecting the positionTop and positionLeft as absolute position. Any chance that it could respect the safearea settings? Or do you know any way to get offset of safearea ? Thanks.

![image](https://github.com/hm21/pro_image_editor/assets/22078631/321ef346-b6c5-4bba-a243-cf544e2f4d51)

**Reply by Chris2137** [reply]
And unfortunately drawing on the final image is not very precise
(draw on the lid)
![image](https://github.com/hm21/pro_image_editor/assets/22078631/be832259-4c3c-4289-9946-5020cb1922df)

shifted downward a bit (but observable)
![image](https://github.com/hm21/pro_image_editor/assets/22078631/43f9a0ae-685d-4943-9641-16634a6313d9)

**Reply by hm21** [reply]
For safearea position you can set for positionTop `MediaQuery.of(context).padding.top`. This also fixes that the position is exactly then, cuz if the subeditor has not same size or position like zhe maineditor, the offset position from the center will also change.

**Reply by Chris2137** [reply]
Got a strange observation, I'm observing a shrinking in image area when in paint editor mode, I tried to pinpoint to the root cause and trying to print the size in buildBody function in main_editor and paint_editor respectively, surprisingly , the size in main_editor is Size(658.0, 584.0), whereas the size in paintEditor is Size(658.0, 540.0), the 44 pixel difference is causing the shifting of drawing in my app. Do you have any idea why is the size difference? 
I even tried to print the height of the appBar and bottomBar, they're both 56 in both view, so I've no idea why the body available height is different in the two views.

**Reply by hm21** [reply]
Sounds interesting. Can you post a minimal example of how I can reproduce it?

**Reply by Chris2137** [reply]
https://github.com/hm21/pro_image_editor/pull/56
Sure thing, it is more observable if you download the code and run in simulator, the image will shrink a bit.
Probably due to the way I calculate the topOffset, but I don't see a proper way to calculate the topOffset. MediaQuery.of(context).padding.top always return 0 in my case.

**Reply by hm21** [reply]
Okay, I have just tested your example. The problem in your code:

```Dart
final topOffset =(MediaQuery.sizeOf(context).height - constraints.maxHeight - 80) / 2;
```
Is that you are calculating the half of the full vertical safearea, which means also the bottom safearea, which will give you a wrong result/position if there is also a navigation bar. Anyway, even if there is no navigation bar it will still not work because I saw the appbar always try to add a safearea by itself, which gives us a wrong position. The easiest way is if I wrap it inside a safearea which should fix your problem and you can set positionTop directly to 0. I release it already in version `3.0.0-dev.15`.

Btw, if you need to get the statusbar height you can do it like below cuz it seems like it no longer works with `MediaQuery.of(context)` on android devices.

```dart
FlutterView view = WidgetsBinding.instance.platformDispatcher.views.first;
EdgeInsets safeArea = MediaQueryData.fromView(view).viewPadding;
```

**Reply by Chris2137** [reply]
Hi @hm21 , I tried with the version 3.0.0-dev.15, it works great. Just a minor issue that, after editing an image, then change the screen orientation, the editor couldn't render correctly. Just for your info. Thanks again for the quick response and fix. I have got another question by the way. I'll start another thread for that.

**Reply by hm21** [reply]
Yes, that's indeed an open task, and it's one of the reasons why it's still in the pre-release stage. You can check out my open tasks [here](https://github.com/users/hm21/projects/5/views/1?filterQuery=) if you curious about it. If I manage to free up enough time this week, I plan to release the final version as well.
